subsys/net/lib/tls_credentials/tls_credentials_trusted.c

```diff
@@ -456,3 +463,238 @@ int tls_credential_delete(sec_tag_t tag, enum tls_credential_type type)
 
 	return ret;
 }
+
+
+#if defined(CONFIG_TLS_CREDENTIAL_CSR)
+/* We need MBEDTLS_USE_PSA_CRYPTO because this implementation uses mbedtls_pk_setup_opaque */
+#if defined(CONFIG_MBEDTLS) && defined(MBEDTLS_USE_PSA_CRYPTO) && defined(MBEDTLS_X509_CSR_WRITE_C)
+
+#include <psa/crypto.h>
+#include "mbedtls/pk.h"
+#include "mbedtls/x509_csr.h"
+
+/**
+ * @brief PSA Random number generator wrapper for Mbed TLS
+ */
+static int psa_rng_for_mbedtls(void *p_rng, unsigned char *output, size_t output_len)
+{
+	ARG_UNUSED(p_rng);
+
+	return psa_generate_random(output, output_len);
+}
+
+/* For this backend, the private-key will be stored in DER format.
+ * Use `cred get <tag> PK bin` to retrieve (the non-terminated base64 encoding of) the key using
+ * the credential shell.
+ */
+int tls_credential_csr(sec_tag_t tag, char *dn, void *csr, size_t *csr_len)
+{
+	int ret = 0;
+
+	int cred_status;
+	psa_status_t psa_status;
+	int mbed_status;
+
+	bool key_created = false;
+	bool key_stored = false;
+
+	/* Credential Storage */
+	psa_storage_uid_t uid = tls_credential_get_uid(tag, TLS_CREDENTIAL_PRIVATE_KEY);
+
+	/* Keygen */
+	psa_key_attributes_t key_attributes = PSA_KEY_ATTRIBUTES_INIT;
+	psa_key_id_t key_id = PSA_KEY_ID_NULL;
+	psa_key_usage_t key_usage;
+
+	/* CSR and key export */
+	mbedtls_pk_context pk_ctx;
+	mbedtls_x509write_csr writer;
+	void* cred_start = 0;
+	size_t csr_max = *csr_len;
```

Comment:
this could be `const size_t csr_max = *csr_len;`

Reply:
Though this would force the declaration to be after the `csr_len` null check, so in fact I cannot if I also follow your suggestion for a `csr_len` NULL check

Reply:
i don't understand.

Reply:
Zephyr generally wants all variable declarations at the beginning of the function.

`const size_t csr_max = *csr_len;` has to happen after the NULL check (unless I do it inline, but that has other issues)

Those two are mutually exclusive

Reply:
zephyr doesn't allow for function input validation?
anyway, the null check is most important.

Reply:
```c
if (csr_len == NULL) return;
const size_t csr_max = *csr_len;
```
is disallowed because declaration is not at start of function

```c
const size_t csr_max;
if (csr_len == NULL) return;
csr_max  = *csr_len; 
```
won't compile.

So I can't add the const keyword